SciML Organization Introduction post (#34)
SciML/sciml.ai: _posts/2020-3-27-SciML.md

world by storm as a interesting blend of traditional scientific mechanistic
modeling (differential equations) with machine learning methodologies like
deep learning. While traditional deep learning methodologies have had difficulties
with interpretability and enforcing physical constraints, this blend with
numerical analysis and differential equations has evolved into a field of research
with new methods, architectures, and algorithms which overcome these traditional
problems. Because of this, new software will be required in order to accommodate
the connection between scientific simulators and scientific machine learning
training loops.

SciML is an open source software organization for the development
and maintenance of a feature-filled and high performance set of tooling for
scientific machine learning. This includes the full gambit of tools from
differential equation solvers to scientific simulators and tools for automatically
discovering the scientific models. What I wanted to do with this post was introduce
the organization by explaining a few things:

- What SciML provides
- What our goals are
- Our next steps
- How you can join in the process

> ## The Service that SciML Provides

Reviewer: Service -> Software?

- [Automatic distributed, multithreaded, and GPU parallelism of ensemble trajectories](https://docs.SciML.org/dev/features/ensemble/)
- [Forward and adjoint local sensitivity analysis](https://docs.SciML.org/dev/analysis/sensitivity/) for fast gradient computations
- Built-in interpolations
- Wraps for common C/Fortran methods like Sundials and Hairer's radau
- Arbitrary precision with BigFloats and Arbfloats
- Arbitrary array types, allowing the definition of differential equations on matrices and distributed arrays

We plan to keep continuing this research indefinitely. Even after
DifferentialEquations.jl tops every benchmark, we plan to add more benchmarks
to find more behaviors and handle those as well. Here is a current benchmark
showcasing native DifferentialEquations.jl methods outperforming classical Fortran
methods like LSODA by 5x on a 20 equation stiff ODE benchmark:

![](https://user-images.githubusercontent.com/1814174/77687352-a0082800-6f74-11ea-924d-442a0836be6d.PNG)
[Reference: Pollution Model Benchmarks](https://benchmarks.sciml.ai/html/StiffODE/Pollution.html)

### We provide tools for deriving and fitting scientific models

> It is very rare that someone things their model is perfect. Instead, a large portion

Reviewer: thinks

Reviewer: Instead doesn't seem to flow well

of the focus of our organization is to help scientific modelers derive equations
and fit models. This includes tools for:

> - [Optimization-based and Bayesian parameter estimation](https://docs.SciML.org/dev/analysis/parameter_estimation/)

Reviewer: Optimizaton-based something? Sounds incomplete

- [Building surrogates of models](https://surrogates.sciml.ai/latest/)
- [Uncertainty quantification](https://docs.SciML.org/dev/analysis/uncertainty_quantification/)

Some of our newer tooling like [DataDrivenDiffEq.jl](https://github.com/SciML/DataDrivenDiffEq.jl)
can even take in timeseries data and generate LaTeX code for the best fitting model
(for a recent demonstration, see [this fitting of a COVID-19 model](https://drive.google.com/file/d/1NxFOtpNHl7oXpdSLM06TEN_oO8QylPYx/view)).

We note that while these tools will continue to be tested with differential
equation models, many of these tools more generally apply scientific models in
general. For example, while our global sensitivity analysis tools have been
documented in the differential equation solver, these methods actually work on
any function `f(p)`:

```julia
using QuasiMonteCarlo, DiffEqSensitivity
function ishi(X)
    A= 7
    B= 0.1
    sin(X[1]) + A*sin(X[2])^2+ B*X[3]^4 *sin(X[1])
end

n = 600000
lb = -ones(4)*π
ub = ones(4)*π
sampler = SobolSample()
A,B = QuasiMonteCarlo.generate_design_matrices(n,lb,ub,sampler)
res1 = gsa(ishi,Sobol(),A,B)
```

> Our reorganization will make it more accessible for users to find our GSA methods

Reviewer: What is GSA?

osol  = solve(oprob, Tsit5())

# solve JumpProblem
u0 = [301, 100, 0, 0]
dprob = DiscreteProblem(rs, u0, tspan, p)
jprob = JumpProblem(dprob, Direct(), rs)
jsol = solve(jprob, SSAStepper())
```

This builds a specific form that can then use optimized methods like `DirectCR`
and achieve an order of magnitude better performance than the classic Gillespie
SSA methods:

![](https://user-images.githubusercontent.com/1814174/77689050-6d136380-6f77-11ea-9248-175de8c1c8e6.PNG)
[Reference: Diffusion Model Benchmarks](https://benchmarks.sciml.ai/html/Jumps/Diffusion_CTRW.html)

Additionally, we have physics-based tooling and support external libraries like:

- [NBodySimulator](https://github.com/SciML/NBodySimulator.jl),
- [RigidBodySim.jl](https://github.com/JuliaRobotics/RigidBodySim.jl) for robotics simulations,
- [QuantumOptics.jl](https://qojulia.org/)
- [DynamicalSystems.jl](https://juliadynamics.github.io/DynamicalSystems.jl/latest/)

We make it easy to generate models of multi-scale systems using tools like
[MultiScaleArrays.jl](https://github.com/SciML/MultiScaleArrays.jl)

![](https://user-images.githubusercontent.com/1814174/27211626-79fe1b9a-520f-11e7-87f1-1cb33da91609.PNG)

> and build compilers like [ModelingToolkit.jl](https://github.com/SciML/ModelingToolkit.jl)

Reviewer: Discuss some underlying optimizations in MTK here?

community. Many of the maintainers of our packages, like Yingbo Ma, Vaibhav Dixit,
Kanav Gupta, Kirill Zubov, etc. all started as one of our over 50 student developers
from past [Google Summer of Code](https://summerofcode.withgoogle.com/) and other
[Julia Seasons of Contributions](https://julialang.org/jsoc/).

## The Goal of SciML

When you read a paper that is
[mixing neural networks with differential equations](https://arxiv.org/abs/2001.04385)
or [designing new neural networks that satisfy incompressibility for modeling Navier-Stokes](https://arxiv.org/abs/1903.00033),
you should be able to go online and find tweakable, high quality, and highly
maintained package implementations of these methodologies to either start using
for your scientific research, or utilize as a starting point for furthering the
methods of scientific machine learning. For this reason, the goal of the SciML
OSS organization is to be a hub for the development of robust cross-language
scientific machine learning software. **In order to make this a reality, we as
an organization have to commit to the following principles**:

### Everything that we build is compatible with automatic differentiation

> Putting an arbitrary piece of code from the SciML group into ae training loop of

Reviewer: ae

---
layout: post
title:  "SciML: An Open Source Software Organization for Scientific Machine Learning"
date:   2020-3-27 10:00:00
categories:
---

Scientific Machine Learning, abbreviated SciML, has been taking the academic

> world by storm as a interesting blend of traditional scientific mechanistic

Reviewer 2: ```suggestion
world by storm as an interesting blend of traditional scientific mechanistic

```

documented in the differential equation solver, these methods actually work on
any function `f(p)`:

```julia
using QuasiMonteCarlo, DiffEqSensitivity
function ishi(X)
    A= 7
    B= 0.1
    sin(X[1]) + A*sin(X[2])^2+ B*X[3]^4 *sin(X[1])
end

n = 600000
lb = -ones(4)*π
ub = ones(4)*π
sampler = SobolSample()
A,B = QuasiMonteCarlo.generate_design_matrices(n,lb,ub,sampler)
res1 = gsa(ishi,Sobol(),A,B)
```

Our reorganization will make it more accessible for users to find out that our
global sensitivity analysis methods to apply in non-differential equation
contexts, like on neural networks.

### We provide high-level domain-specific modeling tools to make scientific modeling more accessible

A physics, biologist, chemist, etc. should be able to pick up our tools and make
use of high performance scientific machine learning methods without requiring the
understanding of every component. To make this a reality, we provide high-level
domain-specific modeling tools as a frontend for building and generating models.
[DiffEqBiological.jl](https://github.com/SciML/DiffEqBiological.jl) is a prime
example which generates high performance simulations from a description of the

> chemical reactions. For example, the following solves the Michaelis-Menton model

Author: Exactly the response I want from this re-branding 👍 .

> chemical reactions. For example, the following solves the Michaelis-Menton model

Reviewer: Would you be renaming some of the packages too?

> chemical reactions. For example, the following solves the Michaelis-Menton model

Author: I need to consider it. DiffEqBiological is a terrible name, DiffEqFlux is mostly not using Flux, and etc. Renaming packages is a little scary though, and I don't like to things that are semi-breaking to users unless they greatly benefit from it.

methods and help create high performance implementations.

In addition, we support many student activities to bring new researchers into the
community. Many of the maintainers of our packages, like Yingbo Ma, Vaibhav Dixit,
Kanav Gupta, Kirill Zubov, etc. all started as one of our over 50 student developers
from past [Google Summer of Code](https://summerofcode.withgoogle.com/) and other
[Julia Seasons of Contributions](https://julialang.org/jsoc/).

## The Goal of SciML

When you read a paper that is
[mixing neural networks with differential equations](https://arxiv.org/abs/2001.04385)
or [designing new neural networks that satisfy incompressibility for modeling Navier-Stokes](https://arxiv.org/abs/1903.00033),
you should be able to go online and find tweakable, high quality, and highly
maintained package implementations of these methodologies to either start using
for your scientific research, or utilize as a starting point for furthering the
methods of scientific machine learning. For this reason, the goal of the SciML
OSS organization is to be a hub for the development of robust cross-language
scientific machine learning software. **In order to make this a reality, we as

> an organization have to commit to the following principles**:

Reviewer 3: ```suggestion
an organization commit to the following principles**:
```

Is this a statement of fact for libraries which can be considered part of SciML, or a call to action?

Author: It's somewhat a statement of fact, but really defining what we've had as a working principle in our head

### Everything that we build is compatible with automatic differentiation

Putting an arbitrary piece of code from the SciML group into a training loop of
some machine learning library like [Flux](https://fluxml.ai/) will naturally work.
This means we plan to enforce coding styles that are compatible with language-wide differentiable programming
tools like [Zygote](https://github.com/FluxML/Zygote.jl), or provide pre-defined

> forward/adjoint rules via derivative rule package [ChainRules.jl](https://github.com/JuliaDiff/ChainRules.jl).

Reviewer 2: ```suggestion
forward/adjoint rules via the derivative rule package [ChainRules.jl](https://github.com/JuliaDiff/ChainRules.jl).

```
assuming you are specifically limiting this to ChainRules

Author: we aren't specifically limiting it to ChainRules, but we hope everything coalesces

As demonstrated in the following animation, you can take our stochastic

> differential equation solvers and training a circuit to control the solution

Reviewer 2: ```suggestion
differential equation solvers and train a circuit to control the solution

```

scientific machine learning to scale, and so we take performance seriously.

### Our packages are routinely and robustly tested with the tools for both scientific simulation and machine learning

This means we will continue to develop tools like
[DiffEqFlux.jl](https://github.com/SciML/DiffEqFlux.jl) which supports the
connection between the [DifferentialEquations.jl](https://docs.sciml.ai/dev/)
differential equation solvers and the [Flux](https://fluxml.ai/) deep learning
library. Other examples this includes our
[surrogate modeling library, Surrogates.jl](https://surrogates.sciml.ai/dev/)
which is routinely tested with DifferentialEquations.jl and the machine learning
AD tooling like Zygote.jl, meaning that you can be sure that our surrogates
modeling tools can train on differential equations and then be used inside
of deep learning stacks. It is this interconnectivity that will allow
next-generation SciML methodologies to get productionized in a way that will
impact "big science" and industrial use.

### We keep up with advances in computational hardware to ensure compatibility with the latest high performance computing tools.

> Today, Intel CPUs and NVIDIA GPUs are the dominant platform, but that won't always

Reviewer 2: ```suggestion
Today, Intel CPUs and NVIDIA GPUs are the dominant platforms, but that won't always

```

be the case. [One of the upcoming top supercomputers will be entirely AMD-based, with AMD CPUs and AMD GPUs](https://www.anandtech.com/show/15581/el-capitan-supercomputer-detailed-amd-cpus-gpus-2-exaflops).
In addition, [Intel GPUs](https://www.anandtech.com/show/15120/intels-2021-exascale-vision-in-aurora-two-sapphire-rapids-cpus-with-six-ponte-vecchio-gpus)

> are scheduled to be a component in the following top supercomputers. We are

Reviewer 2: ```suggestion
are scheduled to be a component in future supercomputers. We are

```
(unless you actually wanted to list some of the "following" supercomputers?)

---
layout: post
title:  "SciML: An Open Source Software Organization for Scientific Machine Learning"
date:   2020-3-27 10:00:00
categories:
---

Scientific Machine Learning, abbreviated SciML, has been taking the academic
world by storm as an interesting blend of traditional scientific mechanistic
modeling (differential equations) with machine learning methodologies like
deep learning. While traditional deep learning methodologies have had difficulties
with interpretability and enforcing physical constraints, this blend with
numerical analysis and differential equations has evolved into a field of research
with new methods, architectures, and algorithms which overcome these traditional
problems. Because of this, new software will be required in order to accommodate
the connection between scientific simulators and scientific machine learning

> training loops.

Reviewer 4: Perhaps it's worth mentioning explicitly in the paragraph above that traditional deep learning approaches usually don't encode prior knowledge that the scientist might have about their system?

SciML is an open source software organization for the development
and maintenance of a feature-filled and high performance set of tooling for

> scientific machine learning. This includes the full gambit of tools from

Reviewer 4: ```suggestion
scientific machine learning. This includes the full gamut of tools from
```
I think gamut is more appropriate

differential equation solvers to scientific simulators and tools for automatically
discovering scientific models. What I wanted to do with this post was introduce

> 

Reviewer 4: ```suggestion
```
don't think there was supposed to be a newline here?

the organization by explaining a few things:

- What SciML provides
- What our goals are
- Our next steps
- How you can join in the process

## The Software that SciML Provides

### We provide best-in-class tooling for solving differential equations

We will continue to have [DifferentialEquations.jl](https://docs.sciml.ai/dev/) at
the core of the organization to support high performance solving of the differential

> 

Reviewer 4: ```suggestion
```
Another unintentional newline?

can even take in timeseries data and generate LaTeX code for the best fitting model
(for a recent demonstration, see [this fitting of a COVID-19 epidemic model](https://drive.google.com/file/d/1NxFOtpNHl7oXpdSLM06TEN_oO8QylPYx/view)).

We note that while these tools will continue to be tested with differential
equation models, many of these tools apply to scientific models in
general. For example, while our global sensitivity analysis tools have been
documented in the differential equation solver, these methods actually work on
any function `f(p)`:

```julia
using QuasiMonteCarlo, DiffEqSensitivity
function ishi(X)
    A= 7
    B= 0.1
    sin(X[1]) + A*sin(X[2])^2+ B*X[3]^4 *sin(X[1])
end

n = 600000
lb = -ones(4)*π
ub = ones(4)*π
sampler = SobolSample()
A,B = QuasiMonteCarlo.generate_design_matrices(n,lb,ub,sampler)
res1 = gsa(ishi,Sobol(),A,B)
```

Reorganizing under the SciML umbrella will make it easier for users to discover
and apply our global sensitivity analysis methods outside of differential equation
contexts, such as with neural networks.

### We provide high-level domain-specific modeling tools to make scientific modeling more accessible

> A physicist, biologist, chemist, etc. should be able to pick up our tools and make

Reviewer 4: ```suggestion
Differential equations appear in nearly every scientific domain, but most
scientific domains have their own specialized idioms and terminology. 
A physicist, biologist, chemist, etc. should be able to pick up our tools and make
```

> 

Reviewer 4: ```suggestion
```

use of high performance scientific machine learning methods without requiring the

> understanding of every component. To make this a reality, we provide high-level

Reviewer 4: ```suggestion
understanding of every component and using abstractions that make sense to
their field. To make this a reality, we provide high-level
```

from as many languages as possible. While there will always be some optimizations
that can only occur when used from the Julia programming language, DSL builders
like [ModelingToolkit.jl](https://github.com/SciML/ModelingToolkit.jl) will be
used to further expand the capabilities and performance of our wrappers.

Here's an example which solves stochastic differential equations with high order adaptive methods

from Python:

```py
# pip install diffeqpy
from diffeqpy import de
# diffeqpy.install()

def f(du,u,p,t):
    x, y, z = u
    sigma, rho, beta = p
    du[0] = sigma * (y - x)
    du[1] = x * (rho - z) - y
    du[2] = x * y - beta * z

def g(du,u,p,t):
    du[0] = 0.3*u[0]
    du[1] = 0.3*u[1]
    du[2] = 0.3*u[2]

numba_f = numba.jit(f)
numba_g = numba.jit(g)
u0 = [1.0,0.0,0.0]
tspan = (0., 100.)
p = [10.0,28.0,2.66]
prob = de.SDEProblem(numba_f, numba_g, u0, tspan, p)
sol = de.solve(prob)

# Now let's draw a phase plot

ut = numpy.transpose(sol.u)
from mpl_toolkits.mplot3d import Axes3D
fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')
ax.plot(ut[0,:],ut[1,:],ut[2,:])
plt.show()
```

> 

Reviewer 4: Maybe mention efforts / plans around static compilation in order to deliver julia interop to other languages with less friction.

and [RootedTrees.jl](https://github.com/SciML/RootedTrees.jl) make it easy to
create and benchmark new methods and accelerate the publication
process for numerical researchers. Our wrappers for external tools like
[FEniCS.jl](https://github.com/SciML/FEniCS.jl) and
[SciPyDiffEq.jl](https://github.com/SciML/SciPyDiffEq.jl) make it easy to perform
cross-platform comparisons. Our stack is entirely written within Julia,
which means every piece can be tweaked on the fly, making it easy to mix
and match Hamiltonian integrators with neural networks to discover new scientific
applications. Our issues and [chat channel](https://gitter.im/JuliaDiffEq/Lobby)
serve as places to not just debug existing software, but also discuss new
methods and help create high performance implementations.

In addition, we support many student activities to bring new researchers into the
community. Many of the maintainers of our packages, like Yingbo Ma, Vaibhav Dixit,
Kanav Gupta, Kirill Zubov, etc. all started as one of our over 50 student developers
from past [Google Summer of Code](https://summerofcode.withgoogle.com/) and other
[Julia Seasons of Contributions](https://julialang.org/jsoc/).

> 

Reviewer 4: Maybe mention that the fact that everything is written in julia makes it easier for students to 'look under the hood' and start understanding and tweaking libraries without learning 5 different languages.

methods of scientific machine learning. For this reason, the goal of the SciML
OSS organization is to be a hub for the development of robust cross-language
scientific machine learning software. **In order to make this a reality, we as
an organization commit to the following principles**:

### Everything that we build is compatible with automatic differentiation

Putting an arbitrary piece of code from the SciML group into a training loop of
some machine learning library like [Flux](https://fluxml.ai/) will naturally work.
This means we plan to enforce coding styles that are compatible with language-wide differentiable programming
tools like [Zygote](https://github.com/FluxML/Zygote.jl), or provide pre-defined
forward/adjoint rules via the derivative rule package [ChainRules.jl](https://github.com/JuliaDiff/ChainRules.jl).

As demonstrated in the following animation, you can take our stochastic
differential equation solvers and train a circuit to control the solution

by simply piecing together compatible packages.

![](https://user-images.githubusercontent.com/1814174/51399524-2c6abf80-1b14-11e9-96ae-0192f7debd03.gif)

> ### Performance is considered a priority, and performance issues are considered a bug

Reviewer 4: ```suggestion
### Performance is considered a priority, and performance issues are considered bugs
```

### Our packages are routinely and robustly tested with the tools for both scientific simulation and machine learning

This means we will continue to develop tools like
[DiffEqFlux.jl](https://github.com/SciML/DiffEqFlux.jl) which supports the
connection between the [DifferentialEquations.jl](https://docs.sciml.ai/dev/)
differential equation solvers and the [Flux](https://fluxml.ai/) deep learning
library. Other examples this includes our
[surrogate modeling library, Surrogates.jl](https://surrogates.sciml.ai/dev/)
which is routinely tested with DifferentialEquations.jl and the machine learning
AD tooling like Zygote.jl, meaning that you can be sure that our surrogates
modeling tools can train on differential equations and then be used inside
of deep learning stacks. It is this interconnectivity that will allow
next-generation SciML methodologies to get productionized in a way that will
impact "big science" and industrial use.

### We keep up with advances in computational hardware to ensure compatibility with the latest high performance computing tools.

Today, Intel CPUs and NVIDIA GPUs are the dominant platforms, but that won't always

> 

Reviewer 4: ```suggestion
```

be the case. [One of the upcoming top supercomputers will be entirely AMD-based, with AMD CPUs and AMD GPUs](https://www.anandtech.com/show/15581/el-capitan-supercomputer-detailed-amd-cpus-gpus-2-exaflops).
In addition, [Intel GPUs](https://www.anandtech.com/show/15120/intels-2021-exascale-vision-in-aurora-two-sapphire-rapids-cpus-with-six-ponte-vecchio-gpus)
are scheduled to be a component in future supercomputers. We are

> 

Reviewer 4: ```suggestion
```

committed to maintaining a SciML toolchain that works on all major platforms,

> updating our compiler backends as new technology is released.

Reviewer 4: Maybe link to https://github.com/jpsamaroo/ROCArrays.jl as a potential avenue.

---
layout: post
title:  "SciML: An Open Source Software Organization for Scientific Machine Learning"
date:   2020-3-27 10:00:00
categories:
---

Computational scientific discovery is at an interesting juncture. While we have
mechanistic models of lots of different scientific phenomena, and reams of data
being generated from experiments - our computational capabilities are unable to

> keep up. Our problems are too large for realistic simulation. Our problems run

Reviewer 4: ```suggestion
keep up. Our problems are too large for realistic simulation. Our problems
```

forward/adjoint rules via the derivative rule package [ChainRules.jl](https://github.com/JuliaDiff/ChainRules.jl).

As demonstrated in the following animation, you can take our stochastic
differential equation solvers and train a circuit to control the solution

by simply piecing together compatible packages.

![](https://user-images.githubusercontent.com/1814174/51399524-2c6abf80-1b14-11e9-96ae-0192f7debd03.gif)

### Performance is considered a priority, and performance issues are considered bugs

No questions asked. If you can find something else that is performing better, we
consider that an issue and should get it fixed. High performance is required for
scientific machine learning to scale, and so we take performance seriously.

### Our packages are routinely and robustly tested with the tools for both scientific simulation and machine learning

This means we will continue to develop tools like
[DiffEqFlux.jl](https://github.com/SciML/DiffEqFlux.jl) which supports the
connection between the [DifferentialEquations.jl](https://docs.sciml.ai/dev/)
differential equation solvers and the [Flux](https://fluxml.ai/) deep learning

> library. Other examples this includes our

Reviewer 5: Should this be:Other examples include ?

scientific machine learning.

[Scientific Machine Learning, abbreviated SciML, has been taking the academic
world by storm as an interesting blend of traditional scientific mechanistic
modeling (differential equations) with machine learning methodologies like
deep learning.](https://www.osti.gov/servlets/purl/1478744) While traditional
deep learning methodologies have had difficulties with scientific issues like
stiffness, interpretability, and enforcing physical constraints, this blend
with numerical analysis and differential equations has evolved into a field of
research with new methods, architectures, and algorithms which overcome these
problems while adding the data-driven automatic learning features of modern
deep learning. Many successes have already been found, with tools like
[physics-informed neural networks](https://www.sciencedirect.com/science/article/pii/S0021999118307125),
[deep BSDE solvers for high dimensional PDEs](https://link.springer.com/article/10.1007%2Fs40304-017-0117-6),
and [neural surrogates](https://arxiv.org/pdf/2001.08055.pdf) showcasing how
deep learning can greatly improve scientific modeling practice. At the same time,

> researcher are quickly finding that our training techniques will need to be

Reviewer: researchers
